src/main/java/org/opensearch/neuralsearch/search/util/NeuralSparseTwoPhaseUtil.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search.util;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BoostQuery;
+import org.apache.lucene.search.Query;
+import org.opensearch.neuralsearch.query.NeuralSparseQuery;
+import org.opensearch.neuralsearch.query.NeuralSparseTwoPhaseParameters;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.rescore.QueryRescorer;
+import org.opensearch.search.rescore.RescoreContext;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.lang.Float.max;
+
+/**
+ * Util class for do two phase preprocess for the NeuralSparseQuery.
+ * Include adding the second phase query to searchContext and set the currentQuery to highScoreTokenQuery.
+ */
+public class NeuralSparseTwoPhaseUtil {
+    /**
+     * @param query         The whole query include neuralSparseQuery to executed.
+     * @param searchContext The searchContext with this query.
+     */
+    public static void addRescoreContextFromNeuralSparseQuery(final Query query, SearchContext searchContext) {
+        Map<Query, Float> query2weight = new HashMap<>();
+        float windowSizeExpansion = populateQueryWeightsMapAndGetWindowSizeExpansion(query, query2weight, 1.0f, 1.0f);
+        Query twoPhaseQuery;
+        if (query2weight.isEmpty()) {
+            return;
+        } else if (query2weight.size() == 1) {
+            Map.Entry<Query, Float> entry = query2weight.entrySet().stream().findFirst().get();
+            twoPhaseQuery = new BoostQuery(entry.getKey(), entry.getValue());
+        } else {
+            twoPhaseQuery = getNestedTwoPhaseQuery(query2weight);
+        }
+        int curWindowSize = (int) (searchContext.size() * windowSizeExpansion);
+        if (curWindowSize < 0 || curWindowSize > NeuralSparseTwoPhaseParameters.MAX_WINDOW_SIZE) {
+            throw new IllegalArgumentException(
+                "Two phase final windowSize "
+                    + curWindowSize
+                    + " out of score with limit "
+                    + NeuralSparseTwoPhaseParameters.MAX_WINDOW_SIZE
+                    + "."
+                    + "You can change the value of cluster setting "
+                    + "[plugins.neural_search.neural_sparse.two_phase.max_window_size] to a integer at least 50."
+            );
+        }
+        QueryRescorer.QueryRescoreContext rescoreContext = new QueryRescorer.QueryRescoreContext(curWindowSize);
+        rescoreContext.setQuery(twoPhaseQuery);
+        rescoreContext.setRescoreQueryWeight(getOriginQueryWeightAfterRescore(searchContext.rescore()));
+        searchContext.addRescore(rescoreContext);
+    }
+
+    private static float populateQueryWeightsMapAndGetWindowSizeExpansion(
```

Comment:
This function needs to be split into 2 parts.
1. Where are are just changing the rescore query/recore context.
2. Where we are changing the actual Query clause provided by user to set what should be the currentQuery that needs to be executed.

We should also add some warnings which should tell us that query clauses was not boost, bool or neuralsparse hence this two phase optimization is not applied.

Reply:
For First comments, I refactor the function addRescoreContextFromNeuralSparseQuery. You can take a look for further optimization.

For the warn logs, I think, If we want to detect all of compound queries that if there has a neuralSparseQuery which couldn't be speed up, maybe it cost some performance loss, such as when a hybrid query, constant score query, dismax query  of a filter boolean query come in, we must to traversal all the sub queries of them and make a warn log.
So I think maybe make a illustration in OpenSearch Document is better for neural-search plugin's performance and log's simplicity.
If you think there is a very strong need to add warn log ?

Reply:
If we only consider the supported compound queires, we can early exit for unsupported cases.
However, if we want to log warn as you mentioned, we need to go through to the end for every compound queries. 
Because here is the common entry for all queries, if users are not using semantic search and need to do many complexed compound queries, going through all of them will have performance impact.

Reply:
I got what you are trying to say. Now help me understand this, if I have a query with multiple level of nesting of bool -> should -> bool -> should -> dis_max -> neuralsparse in this case the optimization will not happen and we will do these iterations to check neural sparse query is present or not.


Moreover now this check is going to happen for every bool query(simple or complex), it has a bigger blast radius and can impact performance of every user query. As the call is recursive so it will iterate over every query clause if it is bool.

Reply:
> I got what you are trying to say. Now help me understand this, if I have a query with multiple level of nesting of bool -> should -> bool -> should -> dis_max -> neuralsparse in this case the optimization will not happen and we will do these iterations to check neural sparse query is present or not.

This is why we need to early exit. We iterate deeper only if its the query clause of specific type. I.e. for a complexed query clause tree, once we found current node is not of specific types we just exit to minimize the performance impact.

With this early exit, we just go through all boolean and boost, and we don't find some use case have deep nested boolean and boost but without nodes we can early exit from.

Reply:
> Another point, if we are not adding any log around whether twoPhaseOptimization got added or not, how we are expecting admin/operators to Opensearch to figure out whether for a given query optimization worked or not?

This feature is not about functionality, but rather an enhancements to search experience. We expect most users can just benefits from the speeds up without learning how it works or configure it(except those who need to customize the settings based on their use case). We'll just pick the part we can speed up and operate it.(like the filter-while-search, we don't have such detailed log there) 

If users want to know whether its enabled, they can just compare the search speed with 2-phase enabled and disabled.

Reply:
> > I got what you are trying to say. Now help me understand this, if I have a query with multiple level of nesting of bool -> should -> bool -> should -> dis_max -> neuralsparse in this case the optimization will not happen and we will do these iterations to check neural sparse query is present or not.
> 
> This is why we need to early exit. We iterate deeper only if its the query clause of specific type. I.e. for a complexed query clause tree, once we found current node is not of specific types we just exit to minimize the performance impact.
> 
> With this early exit, we just go through all boolean and boost, and we don't find some use case have deep nested boolean and boost but without nodes we can early exit from.
> 
> But give an warning log for every query has large impact for performance.

So in my message there 2 points. Seems like you read only one. 

Here is the thing:
1. The code is not going to do any early exit if there is bool clause every where. I am hoping that is clear. 
2. Comparing the latencies of 2 different queries is not an acceptable user explanation here why the queries didn't speed up. So if want operators to understand why a speed up didn't happen we better figure out a way. Otherwise there can be a flurry of questions around why the query is not speeding up. Thats where helpful log message are useful. If there is a better way than log I would suggest to do that.

Reply:
> So in my message there 2 points. Seems like you read only one.
I gave 2 response for the 2 points.

Reply:
The reply for both of them is added. Do you agree with \#1 or not?